RecoBTag/SoftLepton/plugins/SoftPFMuonTagInfoProducer.cc

```diff
@@ -26,132 +33,127 @@
+  const reco::Vertex* vertex=&theVertexCollection->front();
+  
+  // Biult TransientTrackBuilder
+  edm::ESHandle<TransientTrackBuilder> theTrackBuilder;
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder", theTrackBuilder);
+  const TransientTrackBuilder* transientTrackBuilder=theTrackBuilder.product();
+  
+  // Loop on jets
+  for(unsigned int ij=0, nj=theJetCollection->size(); ij<nj; ij++) {
+    edm::RefToBase<reco::Jet> jetRef = theJetCollection->refAt(ij);
+    // Build TagInfo object
+    reco::CandSoftLeptonTagInfo tagInfo;
+    tagInfo.setJetRef(jetRef);
+    // Loop on jet daughters
+    for(unsigned int id=0, nd=jetRef->numberOfDaughters(); id<nd; ++id) {
+      edm::Ptr<reco::Candidate> lepPtr = jetRef->daughterPtr(id);
+      if(abs(lepPtr->pdgId())!=13) continue;
+      
+      const reco::Muon* muon(NULL);
+      // Step 1: try to access the muon from reco::PFCandidate
+      const reco::PFCandidate* pfcand=dynamic_cast<const reco::PFCandidate*>(lepPtr.get());
+      if(pfcand) {
+        muon=pfcand->muonRef().get();
+      }
+      // If not PFCandidate is available, find a match looping on the muon collection
+      else {
+        for(unsigned int im=0, nm=theMuonCollection->size(); im<nm; ++im) { // --- Begin loop on muons
+          const reco::Muon* recomuon=theMuonCollection->refAt(im).get();
+          const pat::Muon* patmuon=dynamic_cast<const pat::Muon*>(recomuon);
+          // Step 2: try a match between reco::Candidate
+          if(patmuon) {
+            if(patmuon->originalObjectRef()==lepPtr) {
+              muon=theMuonCollection->refAt(im).get();
+              break;
+            }
+          }
+          // Step 3: try a match with dR and dpT if pat::Muon casting fails
+          else {
+            if(deltaR(*recomuon, *lepPtr)<0.01 && fabs(recomuon->pt()-lepPtr->pt())/lepPtr->pt()<0.1) {
+              muon=theMuonCollection->refAt(im).get();
+              break;
+            }
+          }
+        } // --- End loop on muons
+      }
+      if(!muon || !muon::isLooseMuon(*muon) || muon->pt()<pTcut) continue;
+      reco::TrackRef trkRef( muon->innerTrack() );
+      reco::TrackBaseRef trkBaseRef( trkRef );
+      
+      // Build Transient Track
+      reco::TransientTrack transientTrack=transientTrackBuilder->build(trkRef);
+      // Calculate variables
+      reco::SoftLeptonProperties properties;
+      properties.sip2d    = IPTools::signedTransverseImpactParameter(transientTrack, GlobalVector(jetRef->px(), jetRef->py(), jetRef->pz()), *vertex).second.significance();
+      properties.sip3d    = IPTools::signedImpactParameter3D(transientTrack, GlobalVector(jetRef->px(), jetRef->py(), jetRef->pz()), *vertex).second.significance();
+      properties.deltaR   = deltaR(*jetRef, *muon);
+      properties.ptRel    = ( (jetRef->p4().Vect()-muon->p4().Vect()).Cross(muon->p4().Vect()) ).R() / jetRef->p4().Vect().R(); // | (Pj-Pu) X Pu | / | Pj |
```

Comment:
On 2/4/15 5:44 AM, zucchett wrote:
> In RecoBTag/SoftLepton/plugins/SoftPFMuonTagInfoProducer.cc
> <https://github.com/cms-sw/cmssw/pull/7542#discussion_r24077702>:
> 
>> +            }
>> +          }
>> +        } // --- End loop on muons
>> +      }
>> +      if(!muon || !muon::isLooseMuon(*muon) || muon->pt()<pTcut) continue;
>> +      reco::TrackRef trkRef( muon->innerTrack() );
>> +      reco::TrackBaseRef trkBaseRef( trkRef );
>> +      
>> +      // Build Transient Track
>> +      reco::TransientTrack transientTrack=transientTrackBuilder->build(trkRef);
>> +      // Calculate variables
>> +      reco::SoftLeptonProperties properties;
>> +      properties.sip2d    = IPTools::signedTransverseImpactParameter(transientTrack, GlobalVector(jetRef->px(), jetRef->py(), jetRef->pz()), *vertex).second.significance();
>> +      properties.sip3d    = IPTools::signedImpactParameter3D(transientTrack, GlobalVector(jetRef->px(), jetRef->py(), jetRef->pz()), *vertex).second.significance();
>> +      properties.deltaR   = deltaR(*jetRef, *muon);
>> +      properties.ptRel    = ( (jetRef->p4().Vect()-muon->p4().Vect()).Cross(muon->p4().Vect()) ).R() / jetRef->p4().Vect().R(); // | (Pj-Pu) X Pu | / | Pj |
> 
> There is no x.Cross(x), but (j-x).Cross(x). The other changes have been
> implemented.

(j - x).Cross(x) = j.Cross(x) - x.Cross(x) = j.Cross(x)

I can imagine you may want j-x instead if ||j-x|| << || j || to have a
better precision.
Is it really the case?


> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/7542/files#r24077702>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------